ogr/ogrsf_frmts/miramon/mm_gdal_functions.c

```diff
@@ -0,0 +1,2801 @@
+            (strcasecmp(data_base_XP->pField[i].FieldName, classical_name)) ==
+                0)
+            return TRUE;
+    }
+    return FALSE;
+}
+
+static char *MM_GiveNewStringWithCharacterInFront(const char *text,
+                                                  char caracter)
+{
+    char *ptr;
+    size_t i;
+
+    if (!text)
+        return nullptr;
+
+    i = strlen(text);
+    if ((ptr = calloc_function(i + 2)) == nullptr)
+        return nullptr;
+
+    *ptr = caracter;
+    memcpy(ptr + 1, text, i + 1);
+    return ptr;
+}
+
+static char *MM_SetSubIndexFieldNam(char *nom_camp, MM_EXT_DBF_N_FIELDS index,
+                                    size_t ampladamax)
+{
+    char *NomCamp_SubIndex;
+    char *_subindex;
+    char subindex[15];
+    size_t sizet_subindex;
+    size_t sizet_nomcamp;
+
+    NomCamp_SubIndex = calloc_function(ampladamax * sizeof(char));
+    if (!NomCamp_SubIndex)
+        return nullptr;
+
+    strcpy(NomCamp_SubIndex, nom_camp);
+
+    sprintf(subindex, sprintf_UINT64, (GUInt64)index);
+
+    _subindex = MM_GiveNewStringWithCharacterInFront(subindex, '_');
+    sizet_subindex = strlen(_subindex);
+    sizet_nomcamp = strlen(NomCamp_SubIndex);
+
+    if (sizet_nomcamp + sizet_subindex > ampladamax - 1)
+        memcpy(NomCamp_SubIndex + ((ampladamax - 1) - sizet_subindex),
+               _subindex, strlen(_subindex));
+    else
+        NomCamp_SubIndex = strcat(NomCamp_SubIndex, _subindex);
+
+    free_function(_subindex);
+
+    return NomCamp_SubIndex;
+}
+
+MM_FIRST_RECORD_OFFSET_TYPE
+MM_GiveOffsetExtendedFieldName(const struct MM_FIELD *camp)
+{
+    MM_FIRST_RECORD_OFFSET_TYPE offset_nom_camp;
+
+    memcpy(&offset_nom_camp,
+           (char *)(&camp->reserved_2) + MM_OFFSET_RESERVAT2_OFFSET_NOM_ESTES,
+           4);
+    return offset_nom_camp;
+}
+
+int MM_WriteNRecordsMMBD_XPFile(struct MMAdmDatabase *MMAdmDB)
+{
+    GUInt32 nRecords;
+    if (!MMAdmDB->pMMBDXP || !MMAdmDB->pFExtDBF)
+        return 0;
+
+    // Updating number of features in features table
+    fseek_function(MMAdmDB->pFExtDBF, MM_FIRST_OFFSET_to_N_RECORDS, SEEK_SET);
+
+    if (MMAdmDB->pMMBDXP->nRecords > UINT32_MAX)
+    {
+        MMAdmDB->pMMBDXP->dbf_version = MM_MARCA_VERSIO_1_DBF_ESTESA;
+
+        if (fwrite_function(&MMAdmDB->pMMBDXP->nRecords, 4, 1,
+                            MMAdmDB->pFExtDBF) != 1)
+            return 1;
+    }
+    else
+    {
+        MMAdmDB->pMMBDXP->dbf_version = MM_MARCA_DBASE4;
+
+        nRecords = (GUInt32)MMAdmDB->pMMBDXP->nRecords;
+        if (fwrite_function(&nRecords, 4, 1, MMAdmDB->pFExtDBF) != 1)
+            return 1;
+    }
+
+    fseek_function(MMAdmDB->pFExtDBF, MM_SECOND_OFFSET_to_N_RECORDS, SEEK_SET);
+    if (MMAdmDB->pMMBDXP->dbf_version == MM_MARCA_VERSIO_1_DBF_ESTESA)
+    {
+        /* from 16 to 19, position MM_SECOND_OFFSET_to_N_RECORDS */
+        if (fwrite_function(((char *)(&MMAdmDB->pMMBDXP->nRecords)) + 4, 4, 1,
```

Comment:
```suggestion
        GUInt32 nRecords32HightBits = (GUInt32)(MMAdmDB->pMMBDXP->nRecords >> 32);
        if (fwrite_function(&nRecords32HightBits, 4, 1,
```

Reply:
I need to be sure you understand what is being done here: the number of records in dBASE IV is a 32 bit integer.
The extension of this format requieres a 64 bit number. Then, all we do is write 4 bytes in the offset MM_FIRST_OFFSET_to_N_RECORDS (the original one) and 4 bytes in the (unused) offset MM_SECOND_OFFSET_to_N_RECORDS (or zero if there if the number needs this 4 extra bytes).

So, you recommend:
The first suggestion is understood. It's kind of transform 64bit number to 32bit number if it doesn't exceed 32bit capacity.
The second one I suppose that our +4 is the >>32.
If it's like that, perfect!

Reply:
Done if no comments. It's on other places of the code (I don't know if you have seens that, but I adopt).
And on the reading part something must be done also.

Reply:
New code:
/* from 4 a 7, position MM_FIRST_OFFSET_to_N_RECORDS */
    if (data_base_XP->nRecords > UINT32_MAX)
    {
        GUInt32 nRecords32HightBits = (GUInt32)(data_base_XP->nRecords >> 32);
        if (fwrite_function(&nRecords32HightBits, 4, 1,
                            data_base_XP->pfDataBase) != 1)
            return FALSE;
    }
    else
    {
        GUInt32 nRecords32LowBits = (GUInt32)(data_base_XP->nRecords & UINT32_MAX);
        if (fwrite_function(&nRecords32LowBits, 4, 1, data_base_XP->pfDataBase) != 1)
            return FALSE;
    }

Reply:
When recomposing the 64bits value:
 if (MM_ES_DBF_ESTESA(pMMBDXP->dbf_version))
    {
        memcpy(&pMMBDXP->nRecords, &nRecords, 4);
        memcpy(((char *)&pMMBDXP->nRecords) **+ 4**, &pMMBDXP->dbf_on_a_LAN, 4);
    }
    else
        pMMBDXP->nRecords = nRecords32LowBits;
        
   Should be used the <<?  (you tell me cause I never use that)

Reply:
yes, on the reading side, this is my remark https://github.com/AbelPau/gdal/pull/12#discussion_r1533746433: "would be cleaner to copy in a nRecordsLow32Bits and nRecordsHigh32Bits, and combine them later with ``nRecords = ((GUInt64)nRecordsHigh32Bits << 32) | nRecordsLow32Bits``. The cast to GUInt64 of the high 32 bits is important to make sure the left bit shift is done correctly"

Reply:
Perfect. DOne